### ISSUE TITLE
False positive ``not-an-iterable`` in match case

### ISSUE THREAD

**Issue by Pierre-Sassoulas** [PyCQA/pylint#5845]
### Bug description

I have 2 functions that do the same thing: recursively iterate an object and return all the string values. I would prefer to use iter2, with the match case, as it is easier on the eye. They both run correctly but pylint reports errors in the second function.

```python
def iter1(object_):
    if isinstance(object_, str):
        yield object_
    if isinstance(object_, list):
        for item in object_:
            yield from iter1(item)
    elif isinstance(object_, dict):
        for value in object_.values():
            yield from iter1(value)
    return

def iter2(object_):
    match object_:
        case str():
            yield object_
        case list():
            for item in object_:
                yield from iter2(item)
        case dict():
            for value in object_.values():
                yield from iter2(value)
        case _:
            return

aaa = {"a": "abc", "b": "123", "c": ["def", "ghi", "jkl"], "d": {"e": "mno", "f": ["pqr", "stu"]}}
print(list(iter1(aaa)))
print(list(iter2(aaa)))
```

See [original question on stackoverflow](https://stackoverflow.com/questions/71284533/pylint-error-not-an-iterable-in-match-case)

### Command used

```shell
pylint a.py
```

### Pylint output

pylint reports these errors in the function iter2.
```shell
Non-iterable value iter2(item) is used in an iterating context pylint (not-an-iterable)
```

### Expected behavior

No error.

### Pylint version

```shell
pylint 2.12.2
astroid 2.9.3
Python 3.10.2 (tags/v3.10.2:a58ebcc, Jan 17 2022, 14:12:15) [MSC v.1929 64 bit (AMD64)]
```

**Comment by jacobtylerwalls** [comment]
Duplicate of #5288
